clippy_lints/src/ref_clone_in_vec_init.rs

```diff
@@ -0,0 +1,81 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::higher::VecArgs;
+use clippy_utils::last_path_segment;
+use clippy_utils::macros::{root_macro_call_first_node, MacroCall};
+use rustc_hir::{Expr, ExprKind, QPath, TyKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{sym, Symbol};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `Arc::new` or `Rc::new` in `vec![elem; len]`
+    ///
+    /// ### Why is this bad?
+    /// This will create `elem` once and clone it `len` times - doing so with `Arc`
+    /// is a bit misleading, as it will create references to the same pointer, rather
+    /// than different `Arc` instances.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let v = vec![std::sync::Arc::new("some data".to_string()); 100];
+    /// // or
+    /// let v = vec![std::rc::Rc::new("some data".to_string())];
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let data = std::sync::Arc::new("some data".to_string());
+    /// // or
+    /// let data = std::rc::Rc::new("some data".to_string());
+    /// let v = vec![data; 100];
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub REF_CLONE_IN_VEC_INIT,
+    suspicious,
+    "initializing `Arc` or `Rc` in `vec![elem; len]`"
+}
+declare_lint_pass!(RefCloneInVecInit => [REF_CLONE_IN_VEC_INIT]);
+
+impl LateLintPass<'_> for RefCloneInVecInit {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        let Some(macro_call) = root_macro_call_first_node(cx, expr) else { return; };
+        let Some(VecArgs::Repeat(elem, _)) = VecArgs::hir(cx, expr) else { return; };
+        let Some(symbol) = new_reference_call(cx, elem) else { return; };
+
+        yield_lint(cx, symbol, &macro_call);
+    }
+}
+
+fn yield_lint(cx: &LateContext<'_>, symbol: Symbol, macro_call: &MacroCall) {
+    let symbol_name = symbol.as_str();
+
+    span_lint_and_then(
+        cx,
+        REF_CLONE_IN_VEC_INIT,
+        macro_call.span,
+        &format!("calling `{symbol_name}::new` in `vec![elem; len]`"),
+        |diag| {
+            diag.note(format!("each `{symbol_name}` will point to the same allocation"));
+            diag.help(format!(
+                "if this is intentional, consider extracting the `{symbol_name}` initialization to a variable"
+            ));
```

Comment:
Here I would add a second help message saying:

`or if not, initilaize each element indivitually`

If you'd like, you can also go one step further and create actual suggestion snippets for each help message with `diag.span_sugg`. The message could look like the ones in the lint description. This is not necessary, but would be a nice addition. I'm also happy to merge this without :upside_down_face:

Reply:
Since its not clear what exactly the user meant (there are 2 options), I think adding these suggestions will make the lint a bit tedious and hard to read
Wdyt?

Reply:
Yeah, this is debatable. I, personally, like to make messages as informative as possible. Both suggestions could also be applied using an IDE. But it can make messages overwhelming. I've created a mockup:

```rs
error: calling `Rc::new` in `vec![elem; len]`
  --> $DIR/simple_case.rs:5:13
   |
LL |     let v = vec![Rc::new("x".to_string()); 2];
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::ref-clone-in-vec-init` implied by `-D warnings`
   = note: each element will point to the same `Rc` instance
help: if this is intentional, consider extracting the `Rc` initialization to a variable
   |
   |    let data = std::rc::Rc::new("some data".to_string());
   |    let v = vec![data; 100];
   |
help: ... or if not, initilaize each element indivitually
   |
   |    let mut v = Vec::with_capacity(100);
   |    for _ in 0..=100 {
   |        v.push(std::rc::Rc::new("some data".to_string()))
   |    }
   |
```

I feel like this is a bit more informative than just the help messages. In rust, I actually focus more on the code snippets and suggestions than the actual message TBH :sweat_smile:. But as said, this is optional :upside_down_face:

Reply:
When I'll get to it I'll try to implement it, lets see how it goes, and how it looks like with a bit more complex snippets 🙂

But since I'm not too familiar with this project, it'd be very kind of you to leave here some references to relevant examples / apis you think i should use to construct these snippets
Thanks ahead 🙏

Reply:
There are several examples in Clippy, you can search for `diag.span_suggestion`. Here is a link to [one example](https://github.com/rust-lang/rust-clippy/blob/98c6381a387ac05e27c03ccfc64146bf4934f5c8/clippy_lints/src/ranges.rs#L368-L388). The basic idea is, that you take the span of the expressions and turn them into strings and format them with a template. So for this, the code could roughly look like this:

```rs
let count_snippet = snippet(cx, len_expr.span, ".."); // `len_expr` can be retrived from `VecArgs::hir(cx, expr)`
let data_snippet = snippet(cx, elem.span, "..");

// And now we just format this:
let mut loop_sugg = "{\n".to_string()
loop_sugg.push(&format!("    let mut v = Vec::with_capacity({count_snippet})\n"));
loop_sugg.push(&format!("    for _ in 0..={count_snippet} {\n"));
loop_sugg.push(&format!("        v.push({data_snippet})\n"));
loop_sugg.push(&format!("    }\n"));
loop_sugg.push(&format!("    v\n"));
loop_sugg.push(&format!("}\n"));
diag.span_suggestion(
    macro_call.span,
    "... or if not, initilaize each element indivitually",
    loop_sugg,
    Applicability::Unspecified,
)
```

this would now suggest replacing the macro with:
```rs
{
    let mut v = Vec::with_capacity(100);
    for _ in 0..=100 {
        v.push(std::rc::Rc::new("some data".to_string()))
    }
    v
}
```

Which would be valid replacement. We could also check if `data_snippet` spans over multiple lines and emit a placeholder if it's too long:

```rs
let mut data_snippet = snippet(cx, elem.span, "..");
if (data_snippet.contains('\n')) {
    // Use a placeholder for multipline element initializers
    data_snippet = "_".to_string();
}
```

(All of this code is untested, so you might have to do some adjustments)

Reply:
Thanks a lot, its more than enough 🙂